### ISSUE TITLE
Implement AMR refiner/derefiner for block systems

### ISSUE THREAD

**Issue by tangqi** [mfem/mfem#818]
Hi, 

We have a working explicit incompressible MHD solver. We would like to examine the AMR capability of mfem. The main thing I am not sure is how to implement an estimator. 

If I understand correctly, the ZZ estimator needs a bilinear integrator that contains two things: `ComputeElementFlux` and `ComputeFluxEnergy`. Rather than implementing those for the system I am solving, I am considering to implement in the following way:
1. Define two estimators. Each will take one block of the solution and compute its flux, similar to https://github.com/mfem/mfem/issues/273. Probably use an integrator similar to `DiffusionIntegrator` at the first pass.
2. Define two refiners and two derefiners accordingly. 
3. In each amr iteration, it will refine twice. At every time step, it will derefine twice. 

This is not very efficient. But since I am not an expert in mfem, that is something I could think of at this stage. Let me know if that makes sense at all, or if there is a better way to do it. 

Thanks,
Qi

**Comment by mlstowell** [comment]
Hello, @tangqi ,

I think you've made a good start but I'd like to offer two alternatives.  As you say, performing the refinement/derefinement twice per time step would not be very efficient.  One way to avoid this would be to combine the results of the two error estimators into a single error measure and pass this to the refiner/derefiner.  The second way is to use the two error estimators to mark elements for refinement but then refine the union of these two sets of elements, similarly derefinement would be done on the intersection of two sets.

If you're following `ex6p.cpp` as a guide, these two options would require that you either define your own `estimator` which uses the results of the two separate ZZ estimators internally to compute a combined error or, for the second option, define your own `refiner` which combines the results of two `ThresholdRefiner` objects. 

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @tangqi ,
On second thought `ex15p.cpp` would be a better guide to follow.  If you choose option 1 and define your own error estimator which combines the results of two existing ZZ error estimators you could pass this to both the `ThresholdRefiner` and `ThresholdDerefiner`.  If you choose option 2, you'll need to implement two of your own classes.

I don't expect either option would be very difficult to implement but I'm not sure which would be more effective.  Perhaps @jakubcerveny would have more insight into that question.

Best,
Mark

**Comment by tangqi** [comment]
@mlstowell Thanks for quick feedback and good suggestions. It seems fairly straightforward to define my own estimator. I have not looked into defining my own refiners, but I would assume it should not be too challenging. Let me know if you have other suggestions @jakubcerveny 

As a nonexpert in fem, I would avoid to define my own integrators if possible, at least at the current stage.

**Comment by jakubcerveny** [comment]
I agree with @mlstowell that you most likely need to merge the results of the two estimators. If you are not able to use the Refiner interface for that, you can always just create a list of elements and call `Mesh::GeneralRefinement`, that's what the Refiner eventually does too. 

Derefinement tends to be more tricky than refinement. One needs to derefine carefully so as not to damage the solution (in contrast, excess refinement is usually not a problem). Also, you are only allowed to derefine certain groups of elements (those that were refined previously). The low-level function for derefinement is `Mesh::DerefineByError`.

**Comment by tangqi** [comment]
Thanks again for the good suggestion @jakubcerveny. Derefiner does look tricky. So I end up writing my own estimator.

BTW, is `anisotropic` in `ZienkiewiczZhuEstimator` a flag for anisotropic AMR? I assume I could do isotropic AMR for now.

**Comment by tzanio** [comment]
> BTW, is anisotropic in ZienkiewiczZhuEstimator a flag for anisotropic AMR? I assume I could do isotropic AMR for now.

Yes to both questions.

**Comment by tangqi** [comment]
Update: My error estimator looks okay. But I have some trouble to update the block vector after mesh is updated.

```c++
   fe_offset[0] = 0;
   fe_offset[1] = fe_size;
   fe_offset[2] = 2*fe_size;
   fe_offset[3] = 3*fe_size;
   fe_offset[4] = 4*fe_size;

   BlockVector vx(fe_offset);
   GridFunction psi, phi, w, j;
   phi.MakeTRef(&fespace, vx.GetBlock(0), 0);
   psi.MakeTRef(&fespace, vx.GetBlock(1), 0);
     w.MakeTRef(&fespace, vx.GetBlock(2), 0);
     j.MakeTRef(&fespace, vx.GetBlock(3), 0);
...
        //in AMR step
        refiner.Apply(*mesh);
        fespace.Update(); //update fem space

       //interpolate grid functions
        psi.Update();
        phi.Update();
          w.Update();
          j.Update();
```
The grid functions looks okay after update. But How could I update the corresponding block vector `vx`? (If the grid functions are reference to a block of `vx`, I do not know what actually happened after `psi.Update()` was called.)

I tried something like `vx.Update(fe_offset)` with an updated `fe_offset`, but I cannot get it right.

**Comment by tangqi** [comment]
OK, I think this should contain the answer: https://github.com/CEED/Laghos/blob/master/amr/laghos.cpp
I will take a look but any hint would be appreciated.

**Comment by mlstowell** [comment]
Hi, @tangqi ,
You may have already found this but have a look at `AMRUpdate()` starting at line 746 of the file you linked to.  I think this is doing something very similar to what you need.

I've been thinking that MFEM may need FEM versions of the `BlockOperator` and `BlockVector` which could handle these sorts of operations.
Best wishes,
Mark

**Comment by tangqi** [comment]
@mlstowell Thanks a lot for the suggestion. Yes, it would be helpful if mfem contains those examples. Currently I need to guess how to use those operators. So here is what I could guess, the interpolations of grid functions happens at
```c++
   const Operator* H1Update = H1FESpace->GetUpdateOperator();
   H1Update->Mult(S_tmp.GetBlock(0), S.GetBlock(0));
```
after the block vector S is created/updated. Is that right?

**Comment by mlstowell** [comment]
Hi, @tangqi ,

Yes, I think you are correct but just for my own benefit let's try to summarize this entire function:
1) Lines 755-756 compute the new numbers of degrees of freedom
2) Lines 761-764 create a new offset array using the new sizes
3) Line 766 Stores a temporary copy of the outdated field data
4) Line 767 Allocates a new array for the field data
5) Lines 769-770 grab the operators that map from the old dofs to the new dofs
6) Lines 772-774 preform the mappings from the outdated field arrays to the newly allocated arrays
7) Lines 776-778 redirect the `GridFunction` objects to their new portions of the `BlockVector`, S.
8) Line 780 resizes the `S_tmp` in preparation for the next update.

Best wishes,
Mark

**Comment by tangqi** [comment]
Hi Mark,
Thanks again for your explanation. That is very helpful for me, probably for other users as well! I will change my code based on your suggestions.
Qi

**Comment by tangqi** [comment]
@mlstowell I have a new question. I write my own `AMRUpdate` based on your suggestion.
```c++
void AMRUpdate(BlockVector &S, BlockVector &S_tmp,
               Array<int> &true_offset,
               GridFunction &phi,
               GridFunction &psi,
               GridFunction &w,
               GridFunction &j)
{
   FiniteElementSpace* H1FESpace = phi.FESpace();

   //update fem space
   H1FESpace->Update();

   int fe_size = H1FESpace->GetTrueVSize();

   //update offset vector
   true_offset[0] = 0;
   true_offset[1] = fe_size;
   true_offset[2] = 2*fe_size;
   true_offset[3] = 3*fe_size;
   true_offset[4] = 4*fe_size;

   S_tmp = S;
   S.Update(true_offset);

   const Operator* H1Update = H1FESpace->GetUpdateOperator();

   H1Update->Mult(S_tmp.GetBlock(0), S.GetBlock(0));
   H1Update->Mult(S_tmp.GetBlock(1), S.GetBlock(1));
   H1Update->Mult(S_tmp.GetBlock(2), S.GetBlock(2));
   H1Update->Mult(S_tmp.GetBlock(3), S.GetBlock(3));

   phi.MakeTRef(H1FESpace, S, true_offset[0]);
   psi.MakeTRef(H1FESpace, S, true_offset[1]);
     w.MakeTRef(H1FESpace, S, true_offset[2]);
     j.MakeTRef(H1FESpace, S, true_offset[3]);

   S_tmp.Update(true_offset);
   H1FESpace->UpdatesFinished();
}
```
I used true dof in my block vector. However, `MakeTRef` cannot link S with grid functions (I checked S is okay). If I follow laghos.cpp and use dof and `MakeRef`, then it works. However, that will mess up my other operators since other operators only apply on true dof.

**Comment by mlstowell** [comment]
Hi, @tangqi ,

Are you sure it's `MakeTRef` that is failing?  If your `BlockVector` is sized according to the true v-size, and you have a nonconforming mesh, then the `H1Update->Mult` lines should fail.  In this case the dimensions of the update operator should be incompatible with the blocks in S and S_tmp.

I'm sure we can come up with a working algorithm but I'd like to first confirm where you're seeing a problem with the above code.

Thanks,
Mark

**Comment by tangqi** [comment]
Hi @mlstowell 
I am not sure what are failing. I can do more tests. (I print S and it looks okay, but the grid functions are just 0 afterwards.) I am fine if I need to switch everything back to dof for my block vector and use `MakeRef`. But I want to follow up with your point:
I noticed thing behaves strangely when my (AMR) mesh is nonconforming. So does `H1Update->Mult` only support conforming meshes?

Qi

**Comment by mlstowell** [comment]
Hi, @tangqi ,

OK, I think that makes sense.  The calls to `MakeTRef` only update an optional vector within a `GridFunction` called `t_vec`.  However, `t_vec` isn't actually used for anything so the important part of `GridFunction` is not being updated.

`H1Update->Mult` will certainly work for either conforming or non-conforming meshes but its size will be different in these two cases so you need to be careful to use vectors with compatible sizes.  Just to be clear, `GridFunction`s have a length that we call `VSize` which accounts for all of the degrees of freedom needed to build the finite element approximations inside each element.  On non-conforming meshes neighboring elements may not share the same degrees of freedom so the `GridFunction` needs to be a little longer to store two different sets of degrees of freedom for a given face.  However, these two different sets of degrees of freedom are related to each other by a set of constraint equations (to ensure the approximation has the correct continuity across the face).  The "true" degree of freedom vector, with size `TrueVSize`, uses these constraint equations to eliminate dependent degrees of freedom so `TrueVSize <= VSize`.

Back to your issue.  If `S` will be based on `TrueVSize` then it isn't large enough, in general, to store the `GridFunction` data so you'll need to store this elsewhere.  This could be another, longer, `BlockVector` or you could just use `phi`, `psi`, `w`, and `j`.  The good news is that I believe we can accomplish this without using `S_tmp`.  Consider the following:
```
/// Let S contain the true dofs associated with phi, psi, w, and j
/// The GridFunctions will be overwritten with data from S
void AMRUpdate(BlockVector &S,
               Array<int> &true_offset,
               GridFunction &phi,
               GridFunction &psi,
               GridFunction &w,
               GridFunction &j)
{
   // Update the GridFunctions so that they match S
   // (this step can be skipped if you can be sure that the GridFunctions already agree with S)
   phi.SetFromTrueDofs(S.GetBlock(0));
   psi.SetFromTrueDofs(S.GetBlock(1));
   w.SetFromTrueDofs(S.GetBlock(2));
   j.SetFromTrueDofs(S.GetBlock(3));

   FiniteElementSpace* H1FESpace = phi.FESpace();

   //update fem space
   H1FESpace->Update();

   // Compute new dofs on the new mesh
   phi.Update();
   psi.Update();
   w.Update();
   j.Update();

   int fe_size = H1FESpace->GetTrueVSize();

   //update offset vector
   true_offset[0] = 0;
   true_offset[1] = fe_size;
   true_offset[2] = 2*fe_size;
   true_offset[3] = 3*fe_size;
   true_offset[4] = 4*fe_size;

   // Resize S
   S.Update(true_offset);

   // Compute "true" dofs and store them in S
   phi.GetTrueDofs(S.GetBlock(0));
   psi.GetTrueDofs(S.GetBlock(1));
     w.GetTrueDofs(S.GetBlock(2));
     j.GetTrueDofs(S.GetBlock(3));

   H1FESpace->UpdatesFinished();
}
```
There are many ways to accomplish these tasks and they each have pros and cons.

Best wishes,
Mark

**Comment by tangqi** [comment]
Hi Mark @mlstowell, 
Thanks again! It works very well. Most AMR parts are working for my block system. 

I have one more question with the operator on the right hand side. My code looks like this:
```c++
  DRe = new BilinearForm(&fespace);
  DRe->AddDomainIntegrator(new DiffusionIntegrator(visc_coeff));  
  DRe->Assemble();  //After this, DRe has a size of fespace.GetTrueVSize()
...
  z.SetSize(fespace.GetTrueVSize());
  DRe->Mult(w, z);  //compute right hand side; note w.Size() is identical to fespace.GetTrueVSize()
... 
  //AMR step and true dof has been changed
...
  AMRUpdate(...);
  DRe->Update();
  DRe->Assemble();  //After this, DRe has a size of fespace.GetVSize()
```
Apparently, it is incompatible. I am trying to fix it. So after every
```c++
  DRe->Assemble();
```
I add
```c++
  DRe.FormSystemMatrix(ess_tdof_list, tmp);
```
Although this fixes the problem, it changes the solution (become wrong). This is very strange to me (I also tested it without AMR.), as I have the impression if a matrix of an operator is not needed (for instance, when it is only applied at the right hand side), I do not need to form a matrix. But when I did form the matrix, it should not effect the solution. So I think I must misunderstand something here.

Qi

**Comment by mlstowell** [comment]
HI, @tangqi ,

I'm glad to hear you're making progress.  When building the RHS I typically work with `GridFunction`s and `BilinearForm`s which should both be sized according to `fespace.GetVSize()`.  We might be able to explain the size of your `DRe` being `fespace.GetTrueVSize()` after the first call to `Assemble()` because your mesh is still conforming (if AMR hasn't begun yet).  Normally `DRe` should have dimensions `fespace.GetVSize()`.

I would suggest that `w` should be a `GridFunction` and `z` can be a `Vector` but it should have a size of `fespace.GetVSize()`.  To make sure this step works you need to make sure that `w` contains the most up-to-date degrees of freedom.  This could be done, for example,  using `w.SetFromTrueDofs(S.GetBlock(2));` as we do in `AMRUpdate()`.  The next thing to do is to prepare `z` to be used as the RHS of a linear solve.  Recall that `z` should have a size of `fespace.GetVsize()` but the RHS should have a size of `fespace.GetTrueVSize()`.  This can be done in at least two ways:
```
   Vector zRHS(fespace.GetTrueVSize());
   const SparseMatrix *P = fespace->GetConformingProlongation();
   P->MultTranspose(z, zRHS);
``` 
or using the `BilinearForm` that creates the matrix being solved:
```
   Vector zRHS;
   a->FormLinearSystem(ess_tdof_list, x, z, A, X, zRHS);
```
The second option is necessary if you're imposing inhomogeneous Dirichlet BCs.  If your BC is zero or if the `ess_tdof_list` is empty then the two options should be equivalent.

Best wishes,
Mark

**Comment by tangqi** [comment]
HI @mlstowell 

Thank you very much for the help. What you said all make sense. This is very helpful (probably should be described somewhere). I am running more tests.

So if I understand you correctly, `DRe->Assemble()` gives an operator of size 
 of `fespace.GetTrueVSize()` if the mesh is conforming, but gives an operator of size `fespace.GetVSize()` if the mesh is nonconforming. Although this seems not ideal, I can see it should be fine if AMR is not used. But since you mentioned it is better to work with operators sized with `fespace.GetVSize()`, for my own curiosity, is there any way to guarantee `DRe->Assemble()` give an operator of correct size even when the mesh is conforming?

PS, now I understand why you store the vectors of size `fespace.GetVSize()` in laghos.

Thanks,
Qi

**Comment by mlstowell** [comment]
HI, @tangqi ,

Actually `DRe->Assemble()` should always produce an operator of size `fespace.GetVSize()`.  However, if the mesh happens to be conforming and it is not partitioned across multiple processors then `fespace.GetVSize()` and `fespace.GetTrueVSize()` return the same number.  If the size of this operator does actually change after calling `Assemble()` when using non-conforming meshes this would be very confusing.  I'm afraid I rarely use the serial versions of these classes so I'm not entirely sure how they function in this situation.

Some of what we've talked about is discussed on the page https://mfem.org/pri-dual-vec/. 

Best wishes,
Mark

**Comment by tangqi** [comment]
Hi @mlstowell 

Thanks again for the help and reference. I will read that more carefully. After reading your comments, I realized it makes more sense to use `fespace.GetVSize()` in the block vector. After changing my code to use `fespace.GetVSize()`, many pieces work well and all the operators on the right hand side using `fespace.GetVSize()` properly. There is only one term left that gives me a small trouble. 

I have a mass/stiffness matrix that is updated using `ess_tdof_list` when inverting it
```c++
M->FormLinearSystem(ess_tdof_list, j, z, A, Y, Z);
```
which will change the size of `M` operator. `M` operator is used on the right hand side later
```c++
M->Mult(w, z);
```
How could I get back to the original operator of the correct size? 

Qi

**Comment by mlstowell** [comment]
Hi, @tangqi ,

I just made a few quick changes to `ex6.cpp` and `ex6p.cpp` and confirmed that the size of a `BilinearForm` object changes after a call to `FormLinearSystem()` but not after a call to `Assemble()`.  In the parallel version the size of the `ParBilinearForm` object does not change.  This is a confusing inconsistency.  I wonder if @v-dobrev or @jakubcerveny would like to offer their opinions on this matter.  I suppose it could be argued that `BilinearForm` is trying to conserve memory by only keeping one version of the matrix but this same argument would apply in the parallel version where two versions of the matrix are stored.

To resolve your current issue I see two options.  First you could build two copies of `M`.  This means more memory and computational effort but remember that applying a BC using `ess_tdof_list` alters select rows of the operator and this would impact the values of `z` on these rows which may not be desirable.  The second approach would be to compute `M->Mult(w,z)` before the call to `FormLinearSystem`.  This may not be possible but consider it.

Best wishes,
Mark

**Comment by tangqi** [comment]
Thanks @mlstowell ! I can live with that for now. I would be interested to know what is the reason behind this. Or maybe I am not using it correctly.

My AMR code starts to run reasonably well. Could I control the level of refinement using `ThresholdRefiner`? I cannot find the option, or is the total number of dof the only option here? (For time dependent problems, it seem reasonable to have minimal grid spacing (levels), since mfem does not do subcycling.)

**Comment by mlstowell** [comment]
Hi, @tangqi ,

@jakubcerveny is the true AMR expert on our team but if I understand your question correctly the `ThresholdRefiner` will only do one level of refinement each time it is called.  However, it is designed to be called within a loop so that you can call it multiple times, for example, until some overall measure of error is achieved, some maximum number of elements or minimal grid spacing is reached , etc..  

Best wishes,
Mark

**Comment by tangqi** [comment]
Thanks for your help, Mark. That was really helpful.

Hi @jakubcerveny I did not find any option to give minimal grid spacing in `ThresholdRefiner`. For a time dependent problem, it would be nice to have it so that I have a good idea of what kind of time step I should use. Is there a better way to use `ThresholdRefiner` to achieve this? Note that I just call refiner once per time step.

Qi

**Comment by jakubcerveny** [comment]
Hi @tangqi, the `ThresholdRefiner` currently does not offer such functionality. It is clear this would be useful though, and we may consider adding a method like `SetMinimumEdgeLength` or `SetMaximumRefinementLevel`. If you'd like to contribute similar improvements to the Refiner, that would be the best :-)

**Comment by tangqi** [comment]
Hi @jakubcerveny, sure, I am happy to contribute. `SetMaximumRefinementLevel` seems not very hard for me to add. But it looks like it would only apply to nonconforming meshes. But I am not sure how levels are assigned to a nonconforming mesh in mfem. For instance, if I print `GetElementDepth` of `data/amr-quad.mesh`, the depth are 2 and 3 for two different meshes.  But I would guess depth would be 0 and 1. Same applies to `fichera-amr.mesh` and other AMR meshes.

On the other hand, it is not clear to me how to add `SetMinimumEdgeLength`. I did not find the function to gives an edge length for each element in `mesh.hpp`.

**Comment by jakubcerveny** [comment]
Determining the maximum refinement level is probably easier. Just be careful to count the levels in all three reference domain axes separately, as `NCMesh` supports anisotropic refinement.

We probably do not have a function to calculate edge length in the Mesh class, but it can definitely be added. It would have to be able to handle both linear and high-order (curved) meshes. The linear case is easy (just calculate the distance between the two vertices), but handling the high-order case correctly means integrating the `Nodes` GridFunction along the edge. For the purpose of AMR an approximation (`GetApproximateEdgeLength`?) could be sufficient, that would only sum the distances between the high order nodes along the edge (these can be obtained by `GetEdgeDofs` and reading `Nodes`).

**Comment by tangqi** [comment]
Hi, @mlstowell @jakubcerveny @v-dobrev, I have a quick question, which I expect has a simple answer. 

The AMR is working (with both refine and derefine) but it refines the region close to boundary which I do not want to refine (it is a bit strange that it is refined there at a later time). So I want to put a constraint based on coordinates of each elements. Which function in `mesh` will give me the node coordinate?

Qi

**Comment by jakubcerveny** [comment]
I think `Mesh::GetNode(i, coord)` should do what you need.

**Comment by tangqi** [comment]
Thanks @jakubcerveny  It works. But the amr code gives me very strange behaviors. I basically change `ThresholdRefiner` to
```c++
   bool yRange=true;
   double vert[2];
   double yMean;

   for (int el = 0; el < NE; el++)
   {
      if (yRange)
      {
        Element *ellocal=mesh.GetElement(el);
        const int nv = ellocal->GetNVertices();
        const int *v = ellocal->GetVertices();
        yMean=0.0;
        for (int j = 0; j < nv; j++)
        {
           mesh.GetNode(v[j], vert);
           yMean+=vert[1];
        }
        yMean=yMean/nv;
        //std::cout <<"el yMean="<<el<<' '<<yMean << '\n';
        //
        if (local_err(el) > threshold && 
            mesh.ncmesh->GetElementDepth(el) < amr_levels &&
            yMean>ymin && yMean<ymax
            )
        {
           marked_elements.Append(Refinement(el));
        }

      }
      else 
      {
        if (local_err(el) > threshold && 
            mesh.ncmesh->GetElementDepth(el) < amr_levels)
        {
           marked_elements.Append(Refinement(el));
        }
      }
   }
```
At the early time, the meshes are identical with or without `yRange`, as I expected. However, the code gives me slightly different solutions (the one with `yRange` has some oscillations outside of [ymin, ymax]). I did not see how that could be possible. Am I missing something here?

**Comment by jakubcerveny** [comment]
I think you need to have `double vert[3]` because `GetNode` always writes three coordinates. The `yMean` variable might be getting overwritten.

**Comment by tangqi** [comment]
Thanks! I tried. But it looks like for meshes where nodes are defined by fem spaces, `GetNode` did not give me the values I want.

```c++
   double vert[3];

   for (int i = 0; i < mesh.GetNE(); i++)
   {
        Element *el=mesh.GetElement(i);
        cout << el->GetGeometryType();
        const int nv = el->GetNVertices();
        const int *v = el->GetVertices();

        for (int j = 0; j < nv; j++)
        {
           cout << ' ' << v[j];
           mesh.GetNode(v[j], vert);
           cout <<" ("<<vert[0]<<','<<vert[1]<<')';
        }
        cout << '\n';
   }
```

Here is the raw output 
3 0 (-1,-1) 1 (-0.333333,-1) 4 (-0.333333,-1) 3 (-0.333333,-0.333333)

My mesh is 
![Screen Shot 2019-04-08 at 1 34 14 PM](https://user-images.githubusercontent.com/1406660/55751523-1c6fc700-5a03-11e9-8587-f0b17bbed2a7.png)

The first element has nodes of 0, 1, 4, 3, which are correctly found, but the coordinates for nodes 3 and 4 are wrong. I will run more tests but any suggestion would be helpful.

**Comment by mlstowell** [comment]
Hi, @tangqi ,
The trouble here is that `Mesh::GetNode()` expects a "node" index but `Element::GetVerticies()` returns "vertex" indices.  Where "Vertices" are topological entities which tie the mesh together at the corners of elements. On the other hand "Nodes" are geometric entities that describe the shape of an element in physical space.

In your mesh the nodes are described using a discontinuous scalar field which has four entries per element which are arranged in a tensor product ordering.  Specifically, the first element contains nodes 0, 1, 2, and 3 where 0 and 1 lie on vertices 0 and 1 in your diagram but node 2 is at vertex 3 and node 3 is at vertex 4.  Your code prints out nodes 0 and 1 correctly because these vertices and nodes happen to be numbered the same.  However, node 4 is actually the first node of the second element which happens to lie on top of the "1" in your diagram.  Similarly, node 3 lies at the location of vertex "4" in your diagram.  I believe this explains your results.

The harder question is how do you actually find the node indices that you want?  I'm not sure about that yet.  I'll try to look into it and post again unless someone else beats me to it.

Best wishes,
Mark

**Comment by mlstowell** [comment]
HI, @tangqi ,
I think the way to approach this is to just treat the nodes as a `GridFunction`, which they are.  This means that you will not know the order of the nodes within a given element but I don't think this will be relevant for your application.  Try this:
```
   double vert[3];

   GridFunction * Nodes = mesh.GetNodes();
   FiniteElementSpace * fes = Nodes->FESpace();
   
   int vdim = fes->GetVDim();
   Array<int> vdofs;
   for (int i = 0; i < mesh.GetNE(); i++)
   {
      fes->GetElementVDofs(i, vdofs);
      int nedofs = vdofs.Size()/vdim;
      for (int j=0; j<nedofs; j++)
      {
         for (int k=0; k<vdim; k++)
         {
            vert[k] = (*Nodes)[vdofs[j+vd*nedofs]];
         }
      }
   }
```
This looks tedious but it makes use of a few assumptions that are at odds with writing a more general function in the Mesh class itself.  One of the most crucial being that we're not interested in the order that we compute these nodes.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Oh, we could actually shorten this a little by using `Mesh::GetNode()` as @jakubcerveny suggested.  Rather than asking for `vdofs` we could just get the `dofs` which are the nodes indices needed by `GetNode()`.  Something like this...
```
   double vert[3];

   GridFunction * Nodes = mesh.GetNodes();
   FiniteElementSpace * fes = Nodes->FESpace();
   
   Array<int> dofs;
   for (int i = 0; i < mesh.GetNE(); i++)
   {
      fes->GetElementDofs(i, dofs);
      for (int j=0; j<dofs.Size(); j++)
      {
        mesh.GetNode(dofs[j], vert);
      }
   }

```

**Comment by tangqi** [comment]
That works well. It is more tricky than I thought. Thanks, Mark. --Qi

**Comment by tangqi** [comment]
I have a mesh reading issue. This is a simple test code to read mesh:
```c++
void PrintLevels(const Mesh *mesh)
{
   for (int el = 0; el < mesh->GetNE(); el++)
   {
      cout<<'('<<el<<','<<mesh->ncmesh->GetElementDepth(el)<<") ";
   }
   cout <<endl;
}

int main(int argc, char *argv[])
{
   const char *mesh_file = "./amr.mesh";
   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.Parse();
   args.PrintOptions(cout);

   Mesh *mesh = new Mesh(mesh_file, 1, 1);
   PrintLevels(mesh);

   delete mesh;
   return 0;
}
```

Made some changes in ex6 to see my issue easier:
```c++
 int main(int argc, char *argv[])
 {
    // 1. Parse command-line options.
@@ -136,7 +145,7 @@ int main(int argc, char *argv[])

    // 11. The main AMR loop. In each iteration we solve the problem on the
    //     current mesh, visualize the solution, and refine the mesh.
-   const int max_dofs = 50000;
+   const int max_dofs = 500;
    for (int it = 0; ; it++)
    {
       int cdofs = fespace.GetTrueVSize();
@@ -198,6 +207,7 @@ int main(int argc, char *argv[])
       //     refined and finally it modifies the mesh. The Stop() method can be
       //     used to determine if a stopping criterion was met.
       refiner.Apply(mesh);
+      PrintLevels(&mesh);
       if (refiner.Stop())
       {
          cout << "Stopping criterion satisfied. Stop." << endl;
@@ -219,5 +229,10 @@ int main(int argc, char *argv[])
       b.Update();
    }

+   ofstream omesh("amr.mesh");
+   PrintLevels(&mesh);
+   omesh.precision(10);
+   mesh.Print(omesh);
+
    return 0;
 }
```

You should be able to reproduce it by ./ex6 -m ../data/star.mesh -o 3
The printed levels are:
(0,0) (1,0) (2,0) (3,0) (4,0) (5,4) (6,4) (7,4) (8,4) (9,3) (10,2) (11,1) (12,0) (13,4) (14,4) (15,4) (16,4) (17,3) (18,2) (19,1) (20,4) (21,4) (22,4) (23,4) (24,3) (25,2) (26,1) (27,0) (28,4) (29,4) (30,4) (31,4) (32,3) (33,2) (34,1) (35,4) (36,4) (37,4) (38,4) (39,3) (40,2) (41,1) (42,0) (43,4) (44,4) (45,4) (46,4) (47,3) (48,2) (49,1) (50,4) (51,4) (52,4) (53,4) (54,3) (55,2) (56,1) (57,0) (58,4) (59,4) (60,4) (61,4) (62,3) (63,2) (64,1) (65,4) (66,4) (67,4) (68,4) (69,3) (70,2) (71,1) (72,0) (73,4) (74,4) (75,4) (76,4) (77,3) (78,2) (79,1)

But when I read mesh, the levels become (which is wrong):
(0,0) (1,0) (2,0) (3,0) (4,0) (5,0) (6,0) (7,0) (8,0) (9,0) (10,4) (11,4) (12,4) (13,4) (14,3) (15,2) (16,1) (17,4) (18,4) (19,4) (20,4) (21,3) (22,2) (23,1) (24,4) (25,4) (26,4) (27,4) (28,3) (29,2) (30,1) (31,4) (32,4) (33,4) (34,4) (35,3) (36,2) (37,1) (38,4) (39,4) (40,4) (41,4) (42,3) (43,2) (44,1) (45,4) (46,4) (47,4) (48,4) (49,3) (50,2) (51,1) (52,4) (53,4) (54,4) (55,4) (56,3) (57,2) (58,1) (59,4) (60,4) (61,4) (62,4) (63,3) (64,2) (65,1) (66,4) (67,4) (68,4) (69,4) (70,3) (71,2) (72,1) (73,4) (74,4) (75,4) (76,4) (77,3) (78,2) (79,1)

The difference is important, as I need it to control my refinement. I also notice sometimes it is identical. So I am not sure if it is because of a bug or because I did not read mesh properly. Thanks!

**Comment by jakubcerveny** [comment]
What is the first line of `amr.mesh`? 

This might be related to a recent change in ordering the elements so that they form a space-filling curve, which is used to partition the mesh in parallel. To be more precise, the orientation of the coarse mesh elements is chosen so that a continuous path exists through the domain, if possible. This will affect the ordering of child (refined) elements. If the mesh is in the old format, your mesh reading example will not do any reorientation. Ex6 will however do it on the first non-conforming refinement. Could you please describe your intended workflow and post the version of your mesh file?

**Comment by tangqi** [comment]
Thanks for the help! amr.mesh is the mesh I saved at the end of ex6 using mesh.Print, not sure if it is in old formate or not. I will check.

My intended workflow is to generate some amr meshes that using fix time iteration, so that I could load and use it at the beginning of time dependent simulation.

The reason is that one operator in my system does not like frequent interpolations (more specifically, I multiply an stiffness matrix with the interpolated solutions and then apply a gradient on it), which will introduce high frequency errors into the sytem. Staring from a precomuted amr grid and refining less frequently should help.

**Comment by tangqi** [comment]
My AMR is working well. But there are some small oscillations appearing in the very late of my simulation. I wish I could do a restart using some AMR mesh and solutions. The above issue prevents me from doing a restart.

So I start to implement a parallel version. Is restart with an AMR mesh and some stored solutions possible for parallel mfem? Restart seems not a feature readily available in mfem. Or do I miss something?

**Comment by jakubcerveny** [comment]
What is the version of the mesh file that you start with? If you start with an `MFEM mesh v1.1` (you can just try increasing it manually), the ordering issue shout not appear.

**Comment by tangqi** [comment]
Do you mean I can just change the first line of star.mesh from 1.0 to 1.1 or do I need to add vertex_parents and so on?

Just changing v1.0 to v1.1 does not seem help...

**Comment by jakubcerveny** [comment]
Ah, you're right, the `vertex_parent` section is not optional. Another thing you can try is to call `mesh->EnsureNCMesh()` always after loading a 1.0 mesh.

**Comment by tangqi** [comment]
I tried various things but it does not work. I can live with it for now. 

I looked in to parallel code. It seems very hard to do a restart with computed solutions and amr meshes. Does anyone have experience on restart? This seems a feature that is necessary for any large scale simultion. An identical restart probably is challenging, but restart with some approximated solutions would be good enough for practical problems.
